azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -1384,4 +1665,161 @@ private void cleanExecutionLogs() {
           - executionLogsRetentionMs);
     }
   }
-}
+
+  /*
+   * This thread is responsible for processing queued flows.
+   */
+  private class QueueProcessorThread extends Thread {
+    private static final long QUEUE_PROCESSOR_WAIT_IN_MS = 1000;
+    private static final int MAX_DISPATCHING_ERRORS_PERMITTED = 5;
+    private final long activeExecutorRefreshWindowInMilisec;
+    private final int activeExecutorRefreshWindowInFlows;
+
+    private boolean shutdown = false;
+    private boolean isActive = true;
+
+    public QueueProcessorThread(boolean isActive,
+      long activeExecutorRefreshWindowInTime,
+      int activeExecutorRefreshWindowInFlows) {
+      setActive(isActive);
+      this.activeExecutorRefreshWindowInFlows =
+        activeExecutorRefreshWindowInFlows;
+      this.activeExecutorRefreshWindowInMilisec =
+        activeExecutorRefreshWindowInTime;
+      this.setName("AzkabanWebServer-QueueProcessor-Thread");
+    }
+
+    public void setActive(boolean isActive) {
+      this.isActive = isActive;
+    }
+
+    public boolean isActive() {
+      return isActive;
+    }
+
+    public void shutdown() {
+      shutdown = true;
+      this.interrupt();
+    }
+
+    public void run() {
+      // Loops till QueueProcessorThread is shutdown
+      while (!shutdown) {
+        synchronized (this) {
+          try {
+            // start processing queue if active, other wait for sometime
+            if (isActive) {
+              processQueuedFlows(activeExecutorRefreshWindowInMilisec,
+                activeExecutorRefreshWindowInFlows);
+            }
+            wait(QUEUE_PROCESSOR_WAIT_IN_MS);
+          } catch (Exception e) {
+            logger.info(
+              "QueueProcessorThread Interrupted. Probably to shut down.", e);
+          }
+        }
+      }
+    }
+
+    /* Method responsible for processing the non-dispatched flows */
+    private void processQueuedFlows(long activeExecutorsRefreshWindow,
+      int maxContinuousFlowProcessed) throws InterruptedException,
+      ExecutorManagerException {
+      long lastExecutorRefreshTime = System.currentTimeMillis();
+      Pair<ExecutionReference, ExecutableFlow> runningCandidate;
+      int currentContinuousFlowProcessed = 0;
+
+      while (isActive() && (runningCandidate = queuedFlows.fetchHead()) != null) {
+        ExecutionReference reference = runningCandidate.getFirst();
+        ExecutableFlow exflow = runningCandidate.getSecond();
+
+        long currentTime = System.currentTimeMillis();
+        if (currentTime - lastExecutorRefreshTime > activeExecutorsRefreshWindow
```

Comment:
As discussed offline,  I have added comment for this if statement. All the variables in this if statements are loop variables and it will be better to have it here instead of having a method with 4 arguments.